### DISCUSSION TITLE
PDF content plus size scaling

### DISCUSSION THREAD

**Discussion by pachhaipurna** [pymupdf/PyMuPDF#3099] [Looking for help] [answered]
@JorjMcKie ,

Hello, I am planning to scale a pdf page and content to specified page height and width,
Is it possible to scale pdf size as well as content, For some pdf size can be decreased whereas for some of the pdf size will e increased, I tried but  I can't figure out.
If there is possible way please help me .

Also, Is it possible to reset he coordinates of pdf, Since for some pdf the origin is in different location.

Thank you.

**Comment by JorjMcKie** [comment]
You can use method `Page.show_pdf_page()` for this.
Assume w, h being the desired page width and height:
```python
import fitz
doc = fitz.open()  # new PDF
src = fitz.open("input.pdf")

for srcpage in src:
    page = doc.new_page(width=w, height=h)
    page.show_page_pdf_page(page.rect, src, srcpage.number)

doc.ez_save("output.pdf")
```
The output pdf will show the original, scaled pages.
Please note:
1. If w, h are not the same aspect ratio (`w/h != src-width / src-height`), there will be empty stripes on the output pages.
3. If input pages are rotated, the code above will need some modification.
4. Links and annotations or fields will not be included in this type of copy. For links at least, there are ways to rescue them.

**Comment by pachhaipurna** [comment]
I am facing issue for note 1

Source page height width and the input provided by us are not matched so
not working properly in that case how can we fixed .

**Reply by pachhaipurna** [reply]
@JorjMcKie for case no 1 how can we fix the problem . Any idea please

**Reply by JorjMcKie** [reply]
It cannot be fixed if you insist on a fixed, file-wide w, h.
You must decide about another approach, like
- accept source page rectangle's aspect ratio but divide each edge by 2 or whatever
- accept that the source page is shown distorted (stretched or shrinked in one dimension)

**Reply by JorjMcKie** [reply]
Please don't be shy and consult the [documentation](https://pymupdf.readthedocs.io/en/latest/page.html#Page.show_pdf_page) of the method!

**Comment by pachhaipurna** [comment]
Could you give me some examples,
So I would be more clear on this .

Thanks

**Reply by JorjMcKie** [reply]
```python
page=doc.new_page(width=300,height=300)
page.show_pdf_page(page.rect,src,0,keep_proportion=False)
44
doc.ez_save("x.pdf")
```
Gives you this
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/b986f41a-bb16-467e-8b14-722274480aad)
from this
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/815808c1-38c6-4986-ae0d-e80041ede7ac)

**Reply by pachhaipurna** [reply]
@JorjMcKie I tried this already, but it is not replacing properly, I am receiving this type of issue.

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/dc1ffaea-8b6b-44da-a388-c426c39d281f)

Here, from the image the blue color is the main pdf page rectangle and yellow color is the new pdf content that is replaced according to new size, the pdf is not for full page according to new pdf size.

The pdf coordinates for origin is in different position.

Can you please let me know how to fix this ?

**Reply by pachhaipurna** [reply]
The content is only added for yellow highlighted color only.

**Comment by JorjMcKie** [comment]
Can you let me have an example PDF and tell me what exactly you want to achieve with it please?

**Reply by pachhaipurna** [reply]
@JorjMcKie The pdf is highly confidential, So I am unable to share it,
 I am extremely sorry for that.
 
 As in my first question I mentioned I want to resize the pdf content height and width .

As in my other program also, I already used this function and this is working fine, However in the pdf that I have currently not is not working .
It is working but some portion of pdf are removed like.

**Reply by pachhaipurna** [reply]
I just want to resize the pdf height and width , All the content must also be resized based on new height and width

**Reply by JorjMcKie** [reply]
Which content is removed? Are pages rotated?

**Reply by JorjMcKie** [reply]
If input pages may be rotated, the code must indeed be adjusted. Ok, another try for what I understood:
1. you want to scale down pages by some factor
2. the width / height ratio must be retained
3. apart from that, pages must look the same
4. the CropBox must equal the MediaBox and start at (0,0)

Everything correct?

**Reply by pachhaipurna** [reply]
@JorjMcKie Yes, totally correct.

**Reply by pachhaipurna** [reply]
@JorjMcKie Is there any process for this?
Thank you.

**Reply by JorjMcKie** [reply, marked as answer]
Ok, good! Alright then.
I am calling the scale factor `scale`. After the following code snippet, pages are resized, have rotation 0 and CropBox = MediaBox.

```python
src = fitz.open("input.pdf")
doc = fitz.open()
for srcpage in src:
    srect = srcpage.rect
    w, h = srect.width, srect.height
    w *= scale
    h *= scale
    rot = srcpage.rotation
    srcpage.set_rotation(0)
    page = doc.new_page(width=w, height=h)
    page.show_pdf_page(page.rect, src, srcpage.number, rotate=-rot)
doc.ez_save("output.pdf")
```

**Reply by pachhaipurna** [reply]
@JorjMcKie 

Thank you so much for your response, It is working correctly, But I have 1 issue.

1.)Once the file height and width are scaled and saved, I need to replace this pdf page with some target page While replacing the annotations must be kept, I already have that one, But once replaced using show_pdf_page the annotations are moved to a different position. Any reason for that?

If I am not wrong annotations are independent of pdf contents.
Any idea on this?

Thank you.

**Reply by JorjMcKie** [reply]
By design, '.show_pdf_page()` **_cannot_** copy annotations or fields.
The main reason being, that the source page is converted to a non-page object, a so-called "Form XObject". This PDF object type cannot have annotations, links or fields - and nothing can be done about this.

With one possible part exception: hyper-links. The script above is basically equivalent to applying a suitably computed `Matrix`. So hyper-links that point to **_external_** resources could be re-created on the target page, IAW GoTo, and Named links will never work.

If you find a way to also recreate annotation on the target, you must of course apply the appropriate scaling, complete transformation. For example, if the source page rectangle is `srect` and that of the (scaled) target is `trect` then you can make a matrix `mat = srect.torect(trect)`. Then the rectangle of an annotation on the target page is `annot.rect * mat`.

**Comment by pachhaipurna** [comment]
As per my understanding target page have markups (funky type annotations)
from bluebeam software , even though I tried to import manually to the file
that we have just generated that annotation are moved in different position
.Any reason for that ?

As per I understand when we use show_pdf_page it must just replace the
target page content . How it is related to the content .

**Reply by JorjMcKie** [reply]
see my comment [here](https://github.com/pymupdf/PyMuPDF/discussions/3099#discussioncomment-8273229).

**Comment by pachhaipurna** [comment]
If we apply transformation for the annotation for funky type that are not
supported by this library , May be it will loose its attribute while doing
transformation.

Any idea to control this ?

**Reply by JorjMcKie** [reply]
see my comment [here](https://github.com/pymupdf/PyMuPDF/discussions/3099#discussioncomment-8273229).

**Reply by pachhaipurna** [reply]
@JorjMcKie I am extremely sorry,I could not understand.

My intention is.

1.) Source page (Generated using scale factor with new height and width). No, annotations.
2) The target page contains annotations having the same height and width as the scaled pdf page.
     
      Need to replace target page content only from the source (Scaled page) , I am receiving the output that annotations are moved in different positions, and doing this logic in the previous (another program) using show_pdf_page working properly.

AS I understand show_pdf_page just replaces content irrespective of annotations.

**Reply by pachhaipurna** [reply]
@JorjMcKie So sorry for asking lots, I have used mat = srect.torect(trect) as per your suggestion,
However when I try to replace target page with transformed coordinates of source page, The original coordinates are saved in the output file.

Here is my code

` mat = src_page.rect.torect(target_page.rect)

    target_page.show_pdf_page(target_page.rect, from_train9_to_new, 0, rotate=0)

    from_train9_to_new.ez_save("output.pdf")`

Do we need to do rotation or not ?

**Reply by JorjMcKie** [reply]
Method `page.show_pdf_page()` replaces nothing on the target page. The source page content is written above the target content.
Existing annotations on the target should not be affected because they are not part of the page itself, but live in a layer above that. So they remain simply visible.

**Reply by JorjMcKie** [reply]
I thought we had been talking about ways to let **_source_** annotations survive???

**Comment by pachhaipurna** [comment]
When I checked source as well as target page what I found is , In source
page the cordinates top left , bottom left, top right and bottom right are
not same as target page . As Target page contains markups ,

When I try to replace target page with source page May be due to
coordinates it replaces the position of markups.

Although the orientation seems same between source and target page and
width height is also same between two pages the top left coordinates
starting are different .

Target page coordinates is (0,0)
Source page coordinates is (0,something)

So, if I can reset the coordinates of source page keeping page orientation
same as target page it can be fixed .

So, I need to know if it’s possible to reset pdf coordinates keeping
orientation same and save it . And just replace to target page .

Can you help me on this ?

**Reply by pachhaipurna** [reply]
@JorjMcKie I solved this issue like this,

`source_pdf = fitz.open(src_pdf)
    src_page =  source_pdf [0]
    rot = src_page.rotation
    src_page.set_rotation(0)

    ##TGT PAGE
    tgt_pdf= fitz.open(tgt_pdf)

    if not(src_page .is_wrapped):
        src_page .wrap_contents()
    ##Clean the pdf contents of page. !!
    src_page .clean_contents(sanitize=True)

    src_page .draw_rect(source_page8.rect,fill=(1,1,1))
    src_page .show_pdf_page(src_page .rect, tgt_pdf, src_page .number, rotate = rot)
    src_page .set_rotation(rot)
 
    source_pdf.ez_save("output.pdf")`

I am not sure if this is correct way to replace source page without changing orientation, even though the starting coordinates between those 2 pdfs are different. I want confirm few things.

1.) clean_contents() will clean all pdf contents before replacing by target pdf.

IS this correct way to replace source page by target page, so when replacing annotations in source page are not rotated and orientation is same,

**Reply by JorjMcKie** [reply]
Well, you never supplied your example PDF's. Therefore, the unprotected geometry changes in the target page couldn't ever be detected.
Wrapping the target contents (or cleaning contents) definitely is a good idea. Not necessary for the source.
Otherwise, congratulations!

**Reply by pachhaipurna** [reply]
@JorjMcKie Thank you so much for your help.
It really works for me and reduce Man hour from 70 percent, I really appreciate for your library.
You are really great. Hope for other help in future also.

Thank you once again.

**Comment by pachhaipurna** [comment]
I am closing this discussion for now because my issue is solved for now.
Thank you.
